db/write_batch_test.cc

```diff
@@ -899,6 +904,89 @@ TEST_F(WriteBatchTest, MemoryLimitTest) {
   ASSERT_TRUE(s.IsMemoryLimit());
 }
 
+TEST_F(WriteBatchTest, AssignTimestamps) {
+  // We assume the last eight bytes of each key is reserved for timestamps.
+  // Therefore, we must make sure each key is longer than eight bytes.
+  constexpr size_t key_size = 16;
+  constexpr size_t num_of_keys = 10;
+  std::vector<std::string> key_strs;
+  for (size_t i = 0; i < num_of_keys; ++i) {
+    key_strs.emplace_back(std::string(key_size, '\0'));
+  }
+  ColumnFamilyHandleImplDummy cf1(0);
+  ColumnFamilyHandleImplDummy cf2(4, test::ComparatorWithU64Ts());
+  ColumnFamilyHandleImplDummy cf3(5, test::ComparatorWithU64Ts());
+
+  WriteBatch batch;
+  // Write to the batch. We will assign timestamps later.
+  for (const auto& key_str : key_strs) {
+    ASSERT_OK(batch.Put(&cf1, key_str, "value"));
+    ASSERT_OK(batch.Put(&cf2, key_str, "value"));
+    ASSERT_OK(batch.Put(&cf3, key_str, "value"));
+  }
+
+  const auto checker1 = [](uint32_t cf, size_t& ts_sz) {
+    if (cf == 4 || cf == 5) {
+      ts_sz = sizeof(uint64_t);
+    } else if (cf == 0) {
+      ts_sz = 0;
+    } else {
+      return Status::Corruption("Invalid cf");
+    }
+    return Status::OK();
+  };
+  ASSERT_OK(
+      batch.AssignTimestamp(std::string(sizeof(uint64_t), '\0'), checker1));
+
+  // We use cf_to_ucmps, cfs_with_ts and timestamp_size to simulate the case
+  // in which a transaction enables indexing for some writes while disables
+  // indexing for other writes. A transaction uses a WriteBatchWithIndex object
+  // to buffer writes (we consider Write-committed policy only). If indexing is
+  // enabled, then writes go through WriteBatchWithIndex API populating a
+  // WBWI internal data structure, i.e. a mapping from cf to user comparators.
+  // If indexing is disabled, a transaction writes directly to the underlying
+  // raw WriteBatch. We will need to track the comparator information for the
+  // column families to which un-indexed writes are performed. When calling
+  // AssignTimestamp(s) API of WriteBatch, we need cf_to_ucmps, cfs_with_ts,
+  // and timestamp_size to perform checking.
+  std::unordered_map<uint32_t, const Comparator*> cf_to_ucmps = {
+      {0, cf1.GetComparator()}, {4, cf2.GetComparator()}};
+  std::unordered_set<uint32_t> cfs_with_ts = {cf3.GetID()};
```

Comment:
IIUC, the default CF and CF 4 play the part of the non-indexed column families here, while CF 5 is indexed. Is that correct? If so, maybe it would be clearer to name these variables accordingly, like `non_indexed_cf_to_ucmps` and `indexed_cfs_with_ts`...

Reply:
Will improve the naming.

Reply:
Done.